README.md
[![Nyan Cat Scoreboard](https://github.com/whatthejeff/nyancat-scoreboard/raw/master/examples.png)](https://github.com/whatthejeff/nyancat-scoreboard/raw/master/examples.png)

## Requirements

The Nyan Cat Scoreboard works with PHP 5.3.3 or later.

## Installation

The recommended way to install the Nyan Cat Scoreboard is
[through composer](http://getcomposer.org). Just create a `composer.json` file
and run the `php composer.phar install` command to install it:

~~~json
{
    "require": {
        "whatthejeff/nyancat-scoreboard": "~1.1"
    }
}
~~~

## Example

~~~php
require_once 'vendor/autoload.php';

use NyanCat\Cat;
use NyanCat\Rainbow;
use NyanCat\Team;
use NyanCat\Scoreboard;

use Fab\SuperFab;

$scoreboard = new Scoreboard(
    new Cat(),
    new Rainbow(
        new SuperFab()
    ),
    array(
        new Team('pass', 'green', '^'),
        new Team('fail', 'red', 'o'),
        new Team('pending', 'cyan', '-'),
    )
);

$scoreboard->start();
for ($i = 0; $i <= 200; $i++) {
    usleep(90000);
    $scoreboard->score('pass');
}
$scoreboard->stop();
~~~

## Tests

[![Build Status](https://travis-ci.org/whatthejeff/nyancat-scoreboard.png?branch=master)](https://travis-ci.org/whatthejeff/nyancat-scoreboard)

To run the test suite, you need [composer](http://getcomposer.org).

    $ php composer.phar install --dev
    $ vendor/bin/phpunit

## License

The Nyan Cat Scoreboard is licensed under the [MIT license](LICENSE).


ChangeLog.md
ChangeLog
=========

This is the list of changes for the Nyan Cat Scoreboard 1.1 release series.

1.1.1
-----

 * Fixed #1: Windows wraps on every line.

1.1.0
-----

* Renamed `NyanCat\Scoreboard::end()` to `NyanCat\Scoreboard::stop()`.
* Added `NyanCat\Scoreboard::isRunning()`.

tests/NyanCat/Tests/CatTest.php
<?php

/*
 * This file is part of Nyan Cat Scoreboard.
 *
 * (c) Jeff Welch <whatthejeff@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace NyanCat\Tests;

use NyanCat\Cat;

/**
 * Cat test cases. (=^.^=)
 *
 * @covers NyanCat\Cat
 * @author Jeff Welch <whatthejeff@gmail.com>
 */
class CatTest extends \PHPUnit_Framework_TestCase
{
    /**
     * An ASCII dinosaur for testing custom states.
     *
     * @var array
     */
    private $dinosaurs = array(
        array(
            '              ___  ',
            '             / %_) ',
            '     _.----./ /    ',
            '    /        /     ',
            ' __/ (  | (  |     ',
            "/__.-'|_|--|_|     "
        ),
        array(
            '               ___ ',
            '              / %_)',
            '     _.----._/ /   ',
            '    /         /    ',
            ' __/ (  | (  |     ',
            "/__.-|_|--|_|      "
        ),
        array(
            '              ___  ',
            '             / %_) ',
            '    _.----._/ /    ',
            '   /         /     ',
            ' _/ (  / /  /      ',
            "/_/-|_/--|_/       "
        )
    );

    /**
     * @dataProvider invalidStatesProvider
     */
    public function testInvalidStates($states, $message)
    {
        $this->setExpectedException('InvalidArgumentException', $message);

        new Cat($states);
    }

    /**
     * @dataProvider validCatProvider
     */
    public function testHeight($states, $height, $width)
    {
        $cat = new Cat($states);
        $this->assertEquals($height, $cat->getHeight());
    }

    /**
     * @dataProvider validCatProvider
     */
    public function testWidth($states, $width, $width)
    {
        $cat = new Cat($states);
        $this->assertEquals($width, $cat->getWidth());
    }

    public function testNext()
    {
        $cat = new Cat($this->dinosaurs);

        $this->assertEquals($this->getDinosaurState(0), $cat->next());
        $this->assertEquals($this->getDinosaurState(1), $cat->next());
        $this->assertEquals($this->getDinosaurState(2), $cat->next());
        $this->assertEquals($this->getDinosaurState(0), $cat->next());
        $this->assertEquals($this->getDinosaurState(1), $cat->next());
        $this->assertEquals($this->getDinosaurState(2), $cat->next());
    }

    public function testNextWithEye()
    {
        $cat = new Cat($this->dinosaurs);

        $this->assertEquals($this->getDinosaurState(0, '^'), $cat->next('^'));
        $this->assertEquals($this->getDinosaurState(1, 'x'), $cat->next('x'));
        $this->assertEquals($this->getDinosaurState(2), $cat->next());
    }

    public function validCatProvider()
    {
        return array(
            array(null, 4, 11),
            array($this->dinosaurs, 6, 19)
        );
    }

    public function invalidStatesProvider()
    {
        return array(
            array(
                array(5),
                'States must be an array of strings'
            ),
            array(
                array(array(5)),
                'States must be an array of strings'
            ),
            array(
                array(array('-'), array('-','_')),
                'Height must be the same for all states'
            )
        );
    }

    private function getDinosaurState($index, $eye = '-')
    {
        return str_replace('%', $eye, $this->dinosaurs[$index]);
    }
}


tests/NyanCat/Tests/Rainbow.php
<?php

/*
 * This file is part of Nyan Cat Scoreboard.
 *
 * (c) Jeff Welch <whatthejeff@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace NyanCat\Tests;

use NyanCat\Rainbow;

/**
 * Rainbow test cases.
 *
 * @covers NyanCat\Rainbow
 * @author Jeff Welch <whatthejeff@gmail.com>
 */
class RainbowTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider invalidRainbowProvider
     */
    public function testInvalidStates($states, $maxWidth, $height, $message)
    {
        $this->setExpectedException('InvalidArgumentException', $message);

        new Rainbow($this->getFab(), $states, $maxWidth, $height);
    }

    /**
     * @dataProvider heightProvider
     */
    public function testHeight($rainbow, $height)
    {
        $this->assertEquals($height, $rainbow->getHeight());
    }

    /**
     * @dataProvider widthProvider
     */
    public function testWidth($rainbow, $width)
    {
        $rainbow->next();
        $this->assertEquals(1, $rainbow->getWidth());

        $rainbow->next();
        $this->assertEquals(2, $rainbow->getWidth());

        for ($i = 0; $i <= $width; $i++) {
            $rainbow->next();
        }

        $rainbow->next();
        $this->assertEquals($width, $rainbow->getWidth());

        $rainbow->next();
        $this->assertEquals($width, $rainbow->getWidth());
    }

    public function testDefaultRainbow()
    {
        $rainbow = new Rainbow($this->getFab());

        $expected = array_fill(0, 4, array('-'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 4, array('-','_'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 4, array('-','_','-'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 4, array('-','_','-','_'));
        $this->assertEquals($expected, $rainbow->next());
    }

    public function testCustomRainbow()
    {
        $rainbow = new Rainbow($this->getFab(), array('*','-','+','/'), 64, 10);

        $expected = array_fill(0, 10, array('*'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 10, array('*','-'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 10, array('*','-','+'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 10, array('*','-','+','/'));
        $this->assertEquals($expected, $rainbow->next());

        $expected = array_fill(0, 10, array('*','-','+','/','*'));
        $this->assertEquals($expected, $rainbow->next());
    }

    public function heightProvider()
    {
        $fab = $this->getFab();

        return array(
            array(
                new Rainbow($fab),
                4
            ),
            array(
                new Rainbow($fab, array('-', '_'), 64, 10),
                10
            ),
        );
    }

    public function widthProvider()
    {
        $fab = $this->getFab();

        return array(
            array(
                new Rainbow($fab),
                64
            ),
            array(
                new Rainbow($fab, array('-', '_'), 100),
                100
            ),
        );
    }

    public function invalidRainbowProvider()
    {
        return array(
            array(
                array(0), 64, 4,
                'State must be a one character string'
            ),
            array(
                array('-_-'), 64, 4,
                'State must be a one character string'
            ),
            array(
                array('-','_'), 'something', 4,
                'Maximum width must be a positive integer'
            ),
            array(
                array('-','_'), 0, 4,
                'Maximum width must be a positive integer'
            ),
            array(
                array('-','_'), 64, 'something',
                'Height must be a positive integer'
            ),
            array(
                array('-','_'), 64, 0,
                'Height must be a positive integer'
            ),
        );
    }

    protected function getFab()
    {
        $fab = $this->getMockBuilder('Fab\SuperFab')
            ->disableOriginalConstructor()
            ->getMock();

        $fab->expects($this->any())
            ->method('paint')
            ->will(
                $this->returnCallback(function ($string) {
                    return $string;
                })
            );

        return $fab;
    }
}


tests/NyanCat/Tests/TeamTest.php
<?php

/*
 * This file is part of Nyan Cat Scoreboard.
 *
 * (c) Jeff Welch <whatthejeff@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace NyanCat\Tests;

use NyanCat\Team;

/**
 * Team test cases.
 *
 * @covers NyanCat\Team
 * @author Jeff Welch <whatthejeff@gmail.com>
 */
class TeamTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider invalidTeamProvider
     */
    public function testInvalidStates($name, $color, $eye, $message)
    {
        $this->setExpectedException('InvalidArgumentException', $message);
        new Team($name, $color, $eye);
    }

    public function testDefaults()
    {
        $team = new Team('test');

        $this->assertEquals(32, $team->getColor());
        $this->assertEquals('-', $team->getEye());
    }

    public function testCustom()
    {
        $team = new Team('test', 'red', '*');

        $this->assertEquals('test', $team->getName());
        $this->assertEquals(31, $team->getColor());
        $this->assertEquals('*', $team->getEye());
    }

    public function testScore()
    {
        $team = new Team('test');
        $this->assertEquals(0, $team->getScore());

        $team->updateScore(10);
        $this->assertEquals(10, $team->getScore());

        $team->updateScore(5);
        $this->assertEquals(15, $team->getScore());

        $team->updateScore(-5);
        $this->assertEquals(10, $team->getScore());
    }

    public function invalidTeamProvider()
    {
        return array(
            array(
                '', 'green', '-',
                'Name must be a non-empty string'
            ),
            array(
                5, 'green', '-',
                'Name must be a non-empty string'
            ),
            array(
                'test', 'Salmon', '-',
                'Invalid foreground color specified: "Salmon"'
            ),
            array(
                'test', 'Salmon', '-',
                'Invalid foreground color specified: "Salmon"'
            ),
            array(
                'test', 'green', '',
                'Eye must be a one character string'
            ),
            array(
                'test', 'green', 'XX',
                'Eye must be a one character string'
            ),
        );
    }
}


src/NyanCat/Cat.php
<?php

/*
 * This file is part of the Nyan Cat Scoreboard.
 *
 * (c) Jeff Welch <whatthejeff@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace NyanCat;

/**
 * ASCII cat representation class.
 *
 * NOTE: Technically this doesn't even need to be a cat, but shhh don't tell.
 * We don't want people knowing they can use this Cat class to represent a
 * dinosaur or something crazy like that. After all, if we wanted to represent
 * a dinosaur, we would need to specify the proper class hierarchy:
 *
 *     Organism > Eukaryote > Animal > Reptile > Dinosaur
 *
 * At least I think that's how this object-oriented stuff works. The wikipedia
 * article is a bit short on examples.
 *
 * @author Jeff Welch <whatthejeff@gmail.com>
 */
class Cat
{
    /**
     * An array of ASCII cat states.
     *
     * Each state should be the array you'd get after running the following on
     * an ASCII cat:
     *
     *     implode("\n", $ascii_state);
     *
     * Use percents (%) for eyes so they can be replaced at certain states to
     * indicate cat emotions.
     *
     * @var array
     */
    private $states = array(
        array(
            '_,------,',
            '_|  /\\_/\\',
            '~|_( % .%)',
            ' ""  ""'
        ),
        array(
            '_,------,',
            '_|   /\\_/\\',
            '^|__( % .%)',
            '  ""  ""'
        )
    );

    /**
     * The number of characters in the widest of the cat states.
     *
     * NOTE: This is calculate automatically.
     *
     * @var integer
     */
    private $width = 11;
    /**
     * The number of newlines in the ASCII cat.
     *
     * NOTE: This is calculate automatically.
     *
     * @var integer
     */
    private $height = 4;

    /**
     * Rolling index for iterating through the cat states.
     *
     * @var integer
     */
    private $index = -1;

    /**
     * Initializes the cat.
     *
     * If you wish to specify a custom ASCII cat, each state should be the
     * array you'd get after running the following on an ASCII cat:
     *
     *     implode("\n", $ascii_state);
     *
     * Use percents (%) for eyes so they can be replaced at certain states to
     * indicate cat emotions.
     *
     * @param array $states An optional array of custom ASCII cat states.
     */
    public function __construct(array $states = null)
    {
        if (empty($states)) {
            return;
        }

        $this->setStates($states);
    }

    /**
     * Gets the number of characters in the widest of the cat states.
     *
     * @return integer The width
     */
    public function getWidth()
    {
        return $this->width;
    }

    /**
     * Gets the number of newlines in the ASCII cat.
     *
     * @return integer The height
     */
    public function getHeight()
    {
        return $this->height;
    }

    /**
     * Gets the next cat state in the rolling index.
     *
     * @param string $eye an optional eye for the next cat state.
     *
     * @return array The next cat state.
     */
    public function next($eye = '-')
    {
        if (!isset($this->states[++$this->index])) {
            $this->index = 0;
        }

        $state = $this->states[$this->index];

        foreach ($state as &$line) {
            $line = str_pad(
                str_replace('%', $eye, $line),
                $this->getWidth()
            );
        }

        return $state;
    }

    /**
     * Sets custom cat states.
     *
     * @param array $states The array of custom ASCII cat states.
     *
     * @throws \InvalidArgumentException When states array is malformed.
     */
    private function setStates(array $states)
    {
        if (empty($states)) {
            throw new \InvalidArgumentException(
                'You must provide cat states'
            );
        }

        $height = count($states[0]);
        $width = 0;

        foreach ($states as $state) {
            if (!is_array($state)) {
                throw new \InvalidArgumentException(
                    'States must be an array of strings'
                );
            }

            if (count($state) !== $height) {
                throw new \InvalidArgumentException(
                    'Height must be the same for all states'
                );
            }

            foreach ($state as $line) {
                if (!is_string($line)) {
                    throw new \InvalidArgumentException(
                        'States must be an array of strings'
                    );
                }

                $width = max($width, strlen($line));
            }
        }

        $this->states = $states;
        $this->width = $width;
        $this->height = $height;
    }
}

src/NyanCat/Team.php
<?php

/*
 * This file is part of the Nyan Cat Scoreboard.
 *
 * (c) Jeff Welch <whatthejeff@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace NyanCat;

/**
 * A team whose score will be represented on the Nyan Cat Scoreboard.
 *
 * @author Jeff Welch <whatthejeff@gmail.com>
 */
class Team
{
    /**
     * ASCII colors.
     *
     * @var array
     */
    private static $colors = array(
        'black'   => 30,
        'red'     => 31,
        'green'   => 32,
        'yellow'  => 33,
        'blue'    => 34,
        'magenta' => 35,
        'cyan'    => 36,
        'white'   => 37
    );

    /**
     * Team name.
     *
     * @var string
     */
    private $name;
    /**
     * Team color.
     *
     * @var integer
     */
    private $color;
    /**
     * Team eye.
     *
     * @var string
     */
    private $eye;

    /**
     * The current team score.
     *
     * @var integer
     */
    private $score = 0;

    /**
     * Initializes the team.
     *
     * @param string $name  The team name.
     * @param string $color The team color.
     * @param string $eye   The team eye.
     */
    public function __construct($name, $color = 'green', $eye = '-')
    {
        $this->setName($name);
        $this->setColor($color);
        $this->setEye($eye);
    }

    /**
     * Gets the team name.
     *
     * @return string The team name.
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * Gets the team color.
     *
     * @return string The team color.
     */
    public function getColor()
    {
        return $this->color;
    }

    /**
     * Gets the team eye.
     *
     * @return string The team eye.
     */
    public function getEye()
    {
        return $this->eye;
    }

    /**
     * Gets the team score.
     *
     * @return integer The team score.
     */
    public function getScore()
    {
        return $this->score;
    }

    /**
     * Updates the team score.
     *
     * @param integer $points The points to add to the team score.
     */
    public function updateScore($points)
    {
        $this->score += $points;
    }

    /**
     * Sets the team color.
     *
     * @param string $color The team color.
     *
     * @throws \InvalidArgumentException When an invalid color is provided.
     */
    private function setColor($color)
    {
        if (!isset(self::$colors[$color])) {
            throw new \InvalidArgumentException(sprintf(
                'Invalid foreground color specified: "%s". Expected one of (%s)',
                $color,
                implode(', ', array_keys(self::$colors))
            ));
        }

        $this->color = self::$colors[$color];
    }

    /**
     * Sets the team name.
     *
     * @param string $name The team name.
     *
     * @throws \InvalidArgumentException When an invalid team name is provided.
     */
    private function setName($name)
    {
        if (!is_string($name) || empty($name)) {
            throw new \InvalidArgumentException(
                'Name must be a non-empty string'
            );
        }

        $this->name = $name;
    }

    /**
     * Sets the team eye.
     *
     * @param string $eye The team eye.
     *
     * @throws \InvalidArgumentException When an invalid team eye is provided.
     */
    private function setEye($eye)
    {
        if (!is_string($eye) || strlen($eye) !== 1) {
            throw new \InvalidArgumentException(
                'Eye must be a one character string'
            );
        }

        $this->eye = $eye;
    }
}

examples/custom-cat.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use NyanCat\Cat;
use NyanCat\Rainbow;
use NyanCat\Team;
use NyanCat\Scoreboard;

use Fab\Factory as FabFactory;

$scoreboard = new Scoreboard(
    new Cat(
        array(
            array(
                '              ___  ',
                '             / %_) ',
                '     _.----./ /    ',
                '    /        /     ',
                ' __/ (  | (  |     ',
                "/__.-'|_|--|_|     "
            ),
            array(
                '               ___ ',
                '              / %_)',
                '     _.----._/ /   ',
                '    /         /    ',
                ' __/ (  | (  |     ',
                "/__.-|_|--|_|      "
            ),
            array(
                '              ___  ',
                '             / %_) ',
                '    _.----._/ /    ',
                '   /         /     ',
                ' _/ (  / /  /      ',
                "/_/-|_/--|_/       "
            )
        )
    ),
    new Rainbow(
        FabFactory::getFab(
            empty($_SERVER['TERM']) ? 'unknown' : $_SERVER['TERM']
        )
    ),
    array(
        new Team('pass', 'green', '^'),
        new Team('fail', 'red', 'o'),
        new Team('pending', 'cyan', '-'),
    )
);

$scoreboard->start();
for($i = 0; $i < 100; $i++) {
    usleep(90000);
    $scoreboard->score('pass');
}
$scoreboard->stop();

src/NyanCat/Scoreboard.php
<?php

/*
 * This file is part of the Nyan Cat Scoreboard.
 *
 * (c) Jeff Welch <whatthejeff@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace NyanCat;

/**
 * The scoreboard where everything is assembled.
 *
 * Example:
 *
 *     use NyanCat\Cat;
 *     use NyanCat\Rainbow;
 *     use NyanCat\Team;
 *     use NyanCat\Scoreboard;
 *
 *     use Fab\SuperFab;
 *
 *     $scoreboard = new Scoreboard(
 *         new Cat(),
 *         new Rainbow(
 *             new SuperFab()
 *         ),
 *         array(
 *             new Team('pass', 'green', '^'),
 *             new Team('fail', 'red', 'o'),
 *             new Team('pending', 'cyan', '-'),
 *         )
 *     );
 *
 *     $scoreboard->start();
 *     for($i = 0; $i <= 200; $i++) {
 *         usleep(90000);
 *         $scoreboard->score('pass');
 *     }
 *     $scoreboard->stop();
 *
 * @author Jeff Welch <whatthejeff@gmail.com>
 */
class Scoreboard
{
    /**
     * Ansi escape
     */
    const ESC = "\x1b[";
    /**
     * Ansi Normal
     */
    const NND = "\x1b[0m";

    /**
     * If a scoreboard is currently running.
     *
     * @var boolean
     */
    private $running = false;

    /**
     * Width in characters of the scores section on the scoreboard.
     *
     * @var integer
     */
    private $width;
    /**
     * Height in newlines of the tallest element on the scoreboard.
     *
     * NOTE: This is set automatically.
     *
     * @var integer
     */
    private $height;

    /**
     * A callable that takes a string and writes it to the appropriate output
     * source. For example:
     *
     *     function ($string) {
     *         print $string;
     *     }
     *
     * @var callable
     */
    private $writer;
    /**
     * The animated ASCII cat.
     *
     * @var NyanCat\Cat
     */
    private $cat;
    /**
     * The animated ASCII rainbow.
     *
     * @var NyanCat\Cat
     */
    private $rainbow;
    /**
     * A collection of teams to track on the scoreboard.
     *
     * @var array
     */
    private $teams;

    /**
     * Initializes the scoreboard.
     *
     * @param Fab\Cat $cat         The animated ASCII cat.
     * @param Fab\Rainbow $rainbow The animated ASCII rainbow.
     * @param array $teams         A collection of teams to track on the scoreboard.
     * @param integer $width       Width in characters of the scores section on the scoreboard.
     * @param callable $writer     A callable that takes a string and writes it to the appropriate output source.
     */
    public function __construct(Cat $cat, Rainbow $rainbow, array $teams, $width = 5, $writer = null)
    {
        $this->cat = $cat;
        $this->rainbow = $rainbow;

        $this->setTeams($teams);
        $this->setWidth($width);
        $this->setHeight();

        $this->setWriter($writer);
    }

    /**
     * Writes out the initial scoreboard.
     */
    public function start()
    {
        $this->running = true;
        $this->score('start');
    }

    /**
     * Updates the score for a team and makes another iteration in the
     * scoreboard animation.
     *
     * @param string $team    The name of the team
     * @param integer $points The number of points to add to the team's score.
     */
    public function score($team, $points = 1)
    {
        if (isset($this->teams[$team])) {
            $this->teams[$team]->updateScore($points);
        }

        $this->drawScoreboard();
        $this->drawRainbow();
        $this->drawCat($team);
    }

    /**
     * Moves the cursor below the scoreboard to end the scoreboard animation.
     */
    public function stop()
    {
        $this->running = false;
        $this->writeLines($this->height);
    }

    /**
     * Checks if a scoreboard is currently running.
     *
     * @return boolean if a scoreboard is running.
     */
    public function isRunning()
    {
        return $this->running;
    }

    /**
     * Gets the number of newlines from the top that it will take to place the
     * current element in the middle of the scoreboard.
     *
     * @param integer $size The size of the current element.
     *
     * @return integer the number of newlines from the top.
     */
    protected function getPreamble($size)
    {
        $height_diff = $this->height - $size;
        return $height_diff > 0 ? floor($height_diff / 2) : 0;
    }

    /**
     * Gets the number of newlines from the end that it will take to place the
     * current element in the middle of the scoreboard.
     *
     * @param integer $size The size of the current element.
     *
     * @return integer the number of newlines from the bottom.
     */
    protected function getPostamble($size)
    {
        $height_diff = $this->height - $size;
        return $height_diff > 0 ? $height_diff - floor($height_diff / 2) : 0;
    }

    /**
     * Prints newlines from the top to place the current element in the middle
     * of the scoreboard.
     *
     * @param integer $size The size of the current element.
     */
    protected function preamble($size)
    {
        $preamble = $this->getPreamble($size);
        if ($preamble > 0) {
            $this->writeLines($preamble);
        }
    }

    /**
     * Prints newlines from the bottom to place the current element in the
     * middle of the scoreboard.
     *
     * @param integer $size The size of the current element.
     */
    protected function postamble($size)
    {
        $postamble = $this->getPostamble($size);
        if ($postamble > 0) {
            $this->writeLines($postamble);
        }

        $this->resetCursor();
    }

    /**
     * Moves the cursor up a number of lines.
     *
     * @param integer $lines The number of lines to move the cursor up.
     */
    protected function cursorUp($lines)
    {
        $this->write(self::ESC . $lines . 'A');
    }

    /**
     * Moves the cursor up to the top of the scoreboard.
     */
    protected function resetCursor()
    {
        $this->cursorUp($this->height);
    }

    /**
     * Moves the cursor down a number of lines.
     *
     * @param integer $lines The number of lines to move the cursor down.
     */
    protected function writeLines($lines)
    {
        if ($lines > 0) {
            $this->write(str_repeat(" \n", $lines));
        }
    }

    /**
     * Writes out a string using the current writer.
     *
     * @param string $string The string to write out.
     */
    protected function write($string)
    {
        call_user_func($this->writer, $string);
    }

    /**
     * Draws the scoreboard for the current iteration.
     */
    protected function drawScoreboard()
    {
        $size = count($this->teams);
        $this->preamble($size);

        foreach ($this->teams as $key => $team) {
            $this->write(sprintf(
                " %s%sm%s%s\n",

                self::ESC,
                $team->getColor(),
                $team->getScore(),
                self::NND
            ));
        }

        $this->postamble($size);
    }

    /**
     * Draws the rainbow for the current iteration.
     */
    protected function drawRainbow()
    {
        $rainbow = $this->rainbow->next();
        $height = $this->rainbow->getHeight();
        $width = $this->rainbow->getWidth();

        $preamble = $this->getPreamble($height);
        $postamble = $this->getPostamble($height);

        while ($preamble-- > 0) {
            $this->write(self::ESC . $this->width . 'C');
            $this->write(str_repeat(' ', $width));
            $this->write("\n");
        }

        foreach ($rainbow as $line) {
            $this->write(self::ESC . $this->width . 'C');
            $this->write(implode($line));
            $this->write("\n");
        }

        while ($postamble-- > 0) {
            $this->write(self::ESC . $this->width . 'C');
            $this->write(str_repeat(' ', $width));
            $this->write("\n");
        }

        $this->resetCursor();

    }

    /**
     * Draws the cat for the current iteration.
     *
     * @param string $team The name of the team that has scored for this iteration.
     */
    protected function drawCat($team)
    {
        $cat = $this->cat->next(
            isset($this->teams[$team]) ? $this->teams[$team]->getEye() : '-'
        );

        $width = $this->width + $this->rainbow->getWidth();
        $size = $this->cat->getHeight();

        $this->preamble($size);

        foreach ($cat as $line) {
            $this->write(self::ESC . $width . 'C');
            $this->write($line);
            $this->write("\n");
        }

        $this->postamble($size);
    }

    /**
     * Sets the teams to track on the scoreboard.
     *
     * @param array $teams the teams to track on the scoreboard.
     *
     * @throws \InvalidArgumentException When invalid or no teams are provided.
     */
    private function setTeams(array $teams)
    {
        if (empty($teams)) {
            throw new \InvalidArgumentException(
                'You must provide at least one team'
            );
        }

        foreach ($teams as $team) {
            if (!$team instanceof Team) {
                throw new \InvalidArgumentException(
                    'All teams must be an instance of NyanCat\\Team'
                );
            }

            $this->teams[$team->getName()] = $team;
        }
    }

    /**
     * Sets the callable that writes out the scoreboard. For example:
     *
     *     function ($string) {
     *         print $string;
     *     }
     *
     * @param callable $writer the callable that writes out the scoreboard.
     *
     * @throws \InvalidArgumentException When $writer is not callable.
     */
    private function setWriter($writer = null)
    {
        if ($writer === null) {
            $writer = function ($string) {
                print $string;
            };
        }

        if (!is_callable($writer)) {
            throw new \InvalidArgumentException(
                'Writer must be callable'
            );
        }

        $this->writer = $writer;
    }

    /**
     * Sets the width in characters of the scores section on the scoreboard.
     *
     * @param integer $width the width in characters.
     *
     * @throws \InvalidArgumentException When $width is not a positive integer.
     */
    private function setWidth($width)
    {
        if (!is_int($width) || $width < 1) {
            throw new \InvalidArgumentException(
                'Width must be a positive interger'
            );
        }

        $this->width = $width;
    }

    /**
     * Sets the height in newlines of the tallest element on the scoreboard.
     */
    private function setHeight()
    {
        $this->height = max(
            count($this->teams),
            $this->rainbow->getHeight(),
            $this->cat->getHeight()
        );
    }
}
